<div style="text-align: center; background-color:#fffff0; padding:10px;">
    <h2 style="font-weight:bold;">Data Preparation</h2>
</div>

**_The key to success in any organization is attracting and retaining top talent. This analysis is useful for an HR Analyst as its task is to determine which factors keep employees at the company and which prompt others to leave. By knowing these factors the HR analyst can change to prevent the loss of good people._**

<div style="text-align: center; background-color: #856ff8; padding:8px;">
    <h2 style: "font-weight: bold;" >OUTLINE</h2>
</div>

- Importing Various Modules
- Loading Dataset
- Data Wrangling
    - Computing the size of DataFrame
    - Enlist the Column Labels
    - Generating Basic Information of Attributes
    - Enlisting Numerical Features
    - Enlisting Categorical Features
    - Checking for Missing Values
    - Descriptive Analysis on Numerical Attributes
    - Drop Unnecessary Columns
    - Descriptive Analysis on Categorical Attributes
    - Checking Unique Values on Categorical Attributes
- Save Dataframe to CSV File

<div style="text-align: center; background-color: yellow; padding:10px;">
    <h2 style="font-weight:bold;">IMPORTING VARIOUS MODULES</h2>
</div>

In [7]:
#Library for Data Manipulation
import pandas as pd
import numpy as np

#Library for Stastical Modelling
from sklearn.preprocessing import LabelEncoder

#Library to ignore warnings
import warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

<div style="text-align: center; background-color: yellow; padding:10px;">
    <h2 style="font-weight:bold;">LOADING DATASET</h2>
</div>

In [9]:
employee_data = pd.read_csv(r"../DATASET/IBM-HR-Analytics-Employee-Attrition-and-Performance.csv")

In [65]:
#print top 5 rows in the dataframe
employee_data.head().style.set_properties(**{'background-color': '#E9F6E2', 'color':'black', 'border-color':'#8b8c8c'})

In [11]:
#print bottom 5 rows of the dataframe
employee_data.tail().style.set_properties(**{'background-color':'#E9F6E2', 'color':'black', 'border-color':'#8b8c8c'})

<div style="text-align:center; background-color:yellow; padding:10px;">
    <h2 style="font-weight:bold;">DATA WRANGLING</h2>
</div>

## <span style="color:blue"> 1] COMPUTING SIZE OF DATASET</span>

In [16]:
#print the shape of the dataframe
print("The shape of the data frame:", employee_data.shape)

#print the length (number of rows) of the data frame
print('Number of rows in the dataframe:', len(employee_data))

#print the number of columns of the dataframe
print('Number od columns in the dataframe:', len(employee_data.columns))

The shape of the data frame: (1470, 35)
Number of rows in the dataframe: 1470
Number od columns in the dataframe: 35


## <span style= 'color:blue;'>2] ENLIST COLUMNS IN DATASET</span>

In [17]:
print("column labels in the dataset in column order:")
for column in employee_data.columns:
    print(column)

column labels in the dataset in column order:
Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EmployeeCount
EmployeeNumber
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
Over18
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StandardHours
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


## <span style='color:blue;'>3] GENERATING BASIC INFORMATION OF ATTRIBUTES</span>

In [20]:
# print the long summary of the dataframe by setting verbose = True
# check for non-Null and Null values in the dataset
employee_data.info(verbose=True, memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

### <font color=red>Inference: </font>
1. There are only 26 Numerical Attributes in the dataset
2. On the other hand we have 9 categorical Attributes

## <span style='color:blue'>4] ENLISTING NUMERICAL FEATURES</span>

In [23]:
employee_data.select_dtypes(include=np.number).sample(5).style.set_properties(**{'background-color':'#E9F6E2', 'color':'black', 'border-color':'#8b8c8c'})

### <font color=red>Inference</font>
1. Some of the numerical features are storing categories in numbers
2. so for better analysis we will replace those labelled numerical values with appropriate values

### 4.1] Labelling Categories in Numerical Features

In [25]:
employee_data['Education'] = employee_data['Education'].replace({1: "Below College", 2: "College", 3:"Bachelor", 4:"Maste", 5:"Doctor"})

In [26]:
employee_data["EnvironmentSatisfaction"] = employee_data["EnvironmentSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [27]:
employee_data["JobInvolvement"] = employee_data["JobInvolvement"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [28]:
employee_data["JobLevel"] = employee_data["JobLevel"].replace({1:"Entry Level",2:"Junior Level",3:"Mid Level",4:"Senior Level",
                                         5:"Executive Level"})

In [29]:
employee_data["JobSatisfaction"] = employee_data["JobSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [30]:
employee_data["PerformanceRating"] = employee_data["PerformanceRating"].replace({1:"Low",2:"Good",3:"Excellent",4:"Outstanding"})

In [31]:
employee_data["RelationshipSatisfaction"] = employee_data["RelationshipSatisfaction"].replace({1:"Low",2:"Medium",3:"High",4:"Very High"})

In [32]:
employee_data["WorkLifeBalance"] = employee_data["WorkLifeBalance"].replace({1:"Bad",2:"Good",3:"Better",4:"Best"})

## <span style='color:blue'>5] ENLISTING CATEGORICAL FEATURES</span>

In [42]:
employee_data.select_dtypes(include='O').sample(5).style.set_properties(**{'background-color': '#E9F6E2',
                                                                'color': 'black','border-color': '#8b8c8c'})

## <span style='color:blue'>6] CHECK FOR MISSING VALUES </span>

In [52]:
missing_df = employee_data.isnull().sum().to_frame().rename(columns={0: "Total No. of Missing Values"})
missing_df["% of Missing Values"] = round((missing_df['Total No. of Missing Values']/len(employee_data) * 100), 2)
missing_df

Total No. of Missing Values  % of Missing Values
Age                                                 0                  0.0
Attrition                                           0                  0.0
BusinessTravel                                      0                  0.0
DailyRate                                           0                  0.0
Department                                          0                  0.0
DistanceFromHome                                    0                  0.0
Education                                           0                  0.0
EducationField                                      0                  0.0
EmployeeCount                                       0                  0.0
EmployeeNumber                                      0                  0.0
EnvironmentSatisfaction                             0                  0.0
Gender                                              0                  0.0
HourlyRate                                          0                  0.0
JobInvolvement                                      0                  0.0
JobLevel                                            0                  0.0
JobRole                                             0                  0.0
JobSatisfaction                                     0                  0.0
MaritalStatus                                       0                  0.0
MonthlyIncome                                       0                  0.0
MonthlyRate                                         0                  0.0
NumCompaniesWorked                                  0                  0.0
Over18                                              0                  0.0
OverTime                                            0                  0.0
PercentSalaryHike                                   0                  0.0
PerformanceRating                                   0                  0.0
RelationshipSatisfaction                            0                  0.0
StandardHours                                       0                  0.0
StockOptionLevel                                    0                  0.0
TotalWorkingYears                                   0                  0.0
TrainingTimesLastYear                               0                  0.0
WorkLifeBalance                                     0                  0.0
YearsAtCompany                                      0                  0.0
YearsInCurrentRole                                  0                  0.0
YearsSinceLastPromotion                             0                  0.0
YearsWithCurrManager                                0                  0.0

### <font color=red> Inference</font>
1. None of the Attributes are having the missing values
2. Since there is no msissing values our further analysis wil be constistant and unbaised.

## <span style='color:blue'>7] DESCRIPTIVE ANALYSIS ON NUMERICAL ATTRIBUTES </span>

In [53]:
employee_data.describe().T

count          mean          std     min      25%  \
Age                      1470.0     36.923810     9.135373    18.0    30.00   
DailyRate                1470.0    802.485714   403.509100   102.0   465.00   
DistanceFromHome         1470.0      9.192517     8.106864     1.0     2.00   
EmployeeCount            1470.0      1.000000     0.000000     1.0     1.00   
EmployeeNumber           1470.0   1024.865306   602.024335     1.0   491.25   
HourlyRate               1470.0     65.891156    20.329428    30.0    48.00   
MonthlyIncome            1470.0   6502.931293  4707.956783  1009.0  2911.00   
MonthlyRate              1470.0  14313.103401  7117.786044  2094.0  8047.00   
NumCompaniesWorked       1470.0      2.693197     2.498009     0.0     1.00   
PercentSalaryHike        1470.0     15.209524     3.659938    11.0    12.00   
StandardHours            1470.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel         1470.0      0.793878     0.852077     0.0     0.00   
TotalWorkingYears        1470.0     11.279592     7.780782     0.0     6.00   
TrainingTimesLastYear    1470.0      2.799320     1.289271     0.0     2.00   
YearsAtCompany           1470.0      7.008163     6.126525     0.0     3.00   
YearsInCurrentRole       1470.0      4.229252     3.623137     0.0     2.00   
YearsSinceLastPromotion  1470.0      2.187755     3.222430     0.0     0.00   
YearsWithCurrManager     1470.0      4.123129     3.568136     0.0     2.00   

                             50%       75%      max  
Age                         36.0     43.00     60.0  
DailyRate                  802.0   1157.00   1499.0  
DistanceFromHome             7.0     14.00     29.0  
EmployeeCount                1.0      1.00      1.0  
EmployeeNumber            1020.5   1555.75   2068.0  
HourlyRate                  66.0     83.75    100.0  
MonthlyIncome             4919.0   8379.00  19999.0  
MonthlyRate              14235.5  20461.50  26999.0  
NumCompaniesWorked           2.0      4.00      9.0  
PercentSalaryHike           14.0     18.00     25.0  
StandardHours               80.0     80.00     80.0  
StockOptionLevel             1.0      1.00      3.0  
TotalWorkingYears           10.0     15.00     40.0  
TrainingTimesLastYear        3.0      3.00      6.0  
YearsAtCompany               5.0      9.00     40.0  
YearsInCurrentRole           3.0      7.00     18.0  
YearsSinceLastPromotion      1.0      3.00     15.0  
YearsWithCurrManager         3.0      7.00     17.0

### <font color=red>Inference:</font>

1. The Minimum Age is 18 which conveys that All employees are Adult. So there's no need of Over18 Attribute for our analysis.
2. The Stanard Deviation value of EmployeeCount and StandardHours is 0.00 which conveys that All values present in this attribute are same.
3. Attribute EmployeeNumber represents a unique value to each of the employees, which will not provide any meaningful inisghts.
4. Since this Attribute will not provide any meaningful insights in our analysis we can simply drop these attributes.

## <span style='color:blue'> 8] DROP UNNECESSARY COLUMNS </span>

**Observation Report:** _We notice that 'EmployeeCount', 'Over18', 'StandardHours' have only one unique values and 'EmployeeNumber' has 1470 unique values. This features aren't useful for us, So we are going to drop those columns._

In [54]:
employee_data.drop(['EmployeeCount', 'Over18', 'StandardHours', 'EmployeeNumber'], axis='columns', inplace=True)

In [55]:
#print top 5 rows in the dataframe
employee_data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome      Education EducationField EnvironmentSatisfaction  \
0                 1        College  Life Sciences                  Medium   
1                 8  Below College  Life Sciences                    High   
2                 2        College          Other               Very High   
3                 3          Maste  Life Sciences               Very High   
4                 2  Below College        Medical                     Low   

   Gender  ...  PerformanceRating RelationshipSatisfaction StockOptionLevel  \
0  Female  ...          Excellent                      Low                0   
1    Male  ...        Outstanding                Very High                1   
2    Male  ...          Excellent                   Medium                0   
3  Female  ...          Excellent                     High                0   
4    Male  ...          Excellent                Very High                1   

  TotalWorkingYears TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                 8                     0             Bad               6   
1                10                     3          Better              10   
2                 7                     3          Better               0   
3                 8                     3          Better               8   
4                 6                     3          Better               2   

   YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
0                   4                        0                    5  
1                   7                        1                    7  
2                   0                        0                    0  
3                   7                        3                    0  
4                   2                        2                    2  

[5 rows x 31 columns]

In [56]:
#print the shape of the dataframe
print("The shape of the data frame:", employee_data.shape)

#print the length (number of rows) of the data frame
print('Number of rows in the dataframe:', len(employee_data))

#print the number of columns of the dataframe
print('Number od columns in the dataframe:', len(employee_data.columns))

The shape of the data frame: (1470, 31)
Number of rows in the dataframe: 1470
Number od columns in the dataframe: 31


In [57]:
print("column labels in the dataset in column order:")
for column in employee_data.columns:
    print(column)

column labels in the dataset in column order:
Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


## <span style='color:blue'> 9] DESCRIPTIVE ANALYSIS ON CATEGORICAL ATTRIBUTES </span>

In [59]:
employee_data.describe(include='O').T

count unique                     top  freq
Attrition                 1470      2                      No  1233
BusinessTravel            1470      3           Travel_Rarely  1043
Department                1470      3  Research & Development   961
Education                 1470      5                Bachelor   572
EducationField            1470      6           Life Sciences   606
EnvironmentSatisfaction   1470      4                    High   453
Gender                    1470      2                    Male   882
JobInvolvement            1470      4                    High   868
JobLevel                  1470      5             Entry Level   543
JobRole                   1470      9         Sales Executive   326
JobSatisfaction           1470      4               Very High   459
MaritalStatus             1470      3                 Married   673
OverTime                  1470      2                      No  1054
PerformanceRating         1470      2               Excellent  1244
RelationshipSatisfaction  1470      4                    High   459
WorkLifeBalance           1470      4                  Better   893

### <font color=red>Inference:</font>

1. All the categorical attributes are having low cardiniality.
2. Attrition and OverTime column is highly biased towards No Category.
3. Businesstravel Attribute is highly biased towards Travel_Rarely category.
4. PerformanceRating Attribute is highly biased towards Excellent category.

## <span style='color:blue'> 10] CHECKING UNIQUE VALUE OF CATEGORICAL ATTRIBUTES </span>

In [63]:
# calculate the number of unique values in each column
for column in employee_data.columns:
    print(f"{column} - Number of Unique values: {employee_data[column].nunique()}")
    print("=============================================================")

Age - Number of Unique values: 43
Attrition - Number of Unique values: 2
BusinessTravel - Number of Unique values: 3
DailyRate - Number of Unique values: 886
Department - Number of Unique values: 3
DistanceFromHome - Number of Unique values: 29
Education - Number of Unique values: 5
EducationField - Number of Unique values: 6
EnvironmentSatisfaction - Number of Unique values: 4
Gender - Number of Unique values: 2
HourlyRate - Number of Unique values: 71
JobInvolvement - Number of Unique values: 4
JobLevel - Number of Unique values: 5
JobRole - Number of Unique values: 9
JobSatisfaction - Number of Unique values: 4
MaritalStatus - Number of Unique values: 3
MonthlyIncome - Number of Unique values: 1349
MonthlyRate - Number of Unique values: 1427
NumCompaniesWorked - Number of Unique values: 10
OverTime - Number of Unique values: 2
PercentSalaryHike - Number of Unique values: 15
PerformanceRating - Number of Unique values: 2
RelationshipSatisfaction - Number of Unique values: 4
StockOpti

In [68]:
categorical_features = []
for column in employee_data.columns:
    if employee_data[column].dtype == object and len(employee_data[column].unique())<= 30:
        categorical_features.append(column)
        print(f"{column} : {employee_data[column].unique()}")
        print(employee_data[column].value_counts())
        print("===========================================================================")
categorical_features.remove("Attrition")
        

Attrition : ['Yes' 'No']
No     1233
Yes     237
Name: Attrition, dtype: int64
BusinessTravel : ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64
Department : ['Sales' 'Research & Development' 'Human Resources']
Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64
Education : ['College' 'Below College' 'Maste' 'Bachelor' 'Doctor']
Bachelor         572
Maste            398
College          282
Below College    170
Doctor            48
Name: Education, dtype: int64
EducationField : ['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64
EnvironmentSatisfaction : ['Medium' 'High' 'Very High' 'Low']
High  